### ISSUE TITLE
Goals

### ISSUE THREAD

**Issue by eed3si9n** [mdedetrich/scala-json-ast#2]
Let's list out the criteria to determine a particular solution/data structure, and try to pick one AST instead of two.
- interop between the common libraries
- performance
  - parsing performance from `String` -> `JValue`
  - pretty printing performance from `JValue -> String`
  - numeric conversion performance from `JNumber` -> `Double` or `Long`
  - field lookup performance from `(JObject, String) -> JValue`
- referential transparency of `String` -> `JValue` -> `String`
- what's stored in `JValue` is guaranteed to produce legal JSON String (this depends on parsing/decoding implementation, not just the choice of types)
-  immutable vs mutable representations
- allow or not allow duplicate keys

If we can say that parsing performance does not matter, I think we could remove `fast`.

**Comment by mdedetrich** [comment]
> referential transparency of String -> JValue -> String

This is really asking for `JObject` to handle duplicate keys (and ordering too), so I am going to lay down what I think of this
1. My attitude to this is similar to the attitude that Linus Torvalds has in this situation. If common practice gets adopted thats out of spec, and (mostly) everyone ends up doing, it does kind of become part of the spec. Even though the JSON spec doesn't put a hardline requirement for not having ordering/duplicate keys, (almost) every language inter-opt (even Javascript!) doesn't handle this. Most languages, when they deal with JSON, want to roughly map the JSON into the appropriate data types that the language holds because this is what users want. This means, for `JObject`, we will be mapping to some sought of Map or Hash, all of which don't handle duplicate keys, and most of which don't handle ordering. The scala Json AST should be the same. That is, when working with `JObject`, we should be working with a `Map`, which means referential transparency goes out the window. They also expect querying to be relatively fast, that is when querying a `JObject` for its key, it should be effectively constant (not possible with an ordered `Map`)
2. Rearding `fast`, it handles two use cases very well. It is designed to be "fast", that is, it uses Arrays as data structures (you need this for parsers implementing their own tricks with Array splicing for performance speed), and it defers all runtime checking and expects parsers to handle this. It also handles the referential transparency/duplicate+ordered keys aspect. If you put in JSON that even has silly stuff like `NaN`, this will be part of `fast`, so it will still hold the String -> JValue -> String referential transparency part
3. The sticking sore thumb in what we are trying to do (relative to other languages) isn't really the strict adherence to the spec, its more the trying to push for referential transparency. I think this is a lost cause when dealing with JSON, especially if you see how loose the spec is and how other languages handle it. This isn't just for `JObject`, but also how the spec treats numbers (no distinction between integers and reals, numbers are also completely unbounded).
4. iirc libraries like Circe/Argonaut aren't referentially transparent according to the spec. They use a `Vector`/`Map` dual data structure for `JObject`, but this is to preserve ordering not for duplicate keys (which isn't possible if you want to provide a Map like interface)

So my kind of final point in this, is that if you want referential transparency in the String -> JValue -> String case, that in my opinion
1. This is an edge case, but obviously people need it
2. This is better served by using a custom structure to handle this, which is analogous to something like Byte Buffers for strings. This is what `fast` is meant to represent, its meant to be "fast" (just like a byte buffer is fast), but its also meant to not lose information due to "validity" reasons, just like a byte buffer does.

If I look at other languages and how they treat JSON (whether static, or dynamic), and I would provide the same implementation in Scala, it essentially would be what `safe` is right now (maybe with some adjustments to how `JNumber` is treated, there isn't any golden solution here though). `Map`/`Vector` is what Scala's own idiomatic data structures for `JArray`/`JObject` are.

So tl;dr version, due to how the JSON spec is designed and how general users expect to treat JSON, I don't see how we can achieve those goals while still being referentially transparent. There may be some solution, but its very likely to be not nice at all, and would probably fall outside of scope of what we want in a Scala module (i.e to be minimal and clean and easy to maintain)

**Comment by mdedetrich** [comment]
Also to be more specific to your points

> interop between the common libraries

I think this is already a given, unless I am missing something

> parsing performance from String -> JValue

Generally speaking, I think people expect that safety comes at a performance cost. So if you want to be fast, you usually do need to be unsafe. 

> pretty printing performance from JValue -> String

I don't believe there is a common need for fast pretty printing, usually if you need fast pretty printing you will probably have the `String` or `Stream` already available before you try and turn it into JSON. If the `String`/`Stream` is not pretty printed already, there is probably another library that fills that use case better. I don't think we should worry about this at all

> numeric conversion performance from JNumber -> Double or Long

This one is a to work with. The fact that Scala doesn't have a proper unbounded lazy real type makes it even worse. Double's are faster than everything, but them being Doubles means they can't represent things like `0.1` or equality properly. `BigDecimal'`s offer a nice compromise for having a precise representation of number, but a `BigDecimal` exposes methods which are very expensive, and they are bounded.

To be clear, `BigDecimal` is only really slow when you do actual math on it. If you are just using it to store numbers its not that small (and the runtime construction of a `BigDecimal` isn't that bad either considering its also validating).

This then brings us to the final point of whether we just store the number as a `String` and use something like a regex to validate if its correct number.

> field lookup performance from (JObject, String) -> JValue

I think that something that users don't really have a tolerance for, is that when querying JSON they have to deal with a data structure that has terrible lookup characteristics. So if we have a single AST, I don't believe that using anything that is not backed by a `Map` for a `JObject` or a `Array`/`Vector` for a JArray is going to do us any good.

> referential transparency of String -> JValue -> String

I think this is an edge case, and it should be something thats solved either internally or with `unsafe.JValue` (`unsafe.JValue` also tries to solve for performance, it just happens to handle RT as well).

Outside of Scala, from what I see, the majority of JSON AST's use a HashMap (mutable or immutable representation) to store a `JObject`. This includes
- Haskell
- Rust
- Python
- Ruby
- Javascript

There are some exceptions, apparently Julia uses a dual backed structure (map + array to keep track of insertion order), however a thing to note here is that mutable data structures use a lot less memory than their immutable counterparts. `Map` + `Vector` does take up a non trivial amount of more memory compared to `Map` + `Array`. I believe that for the idiomatic AST (which is currently `JValue`) people would like to use immutable data structures. You could use an immutable interface to a mutable data structure, but then you lose out on structural sharing which sends the wrong to people.

> what's stored in JValue is guaranteed to produce legal JSON String (this depends on parsing/decoding implementation, not just the choice of types)

So the current issue is that the JSON spec is vague in some areas. i.e. for things like `JObject`, key ordering and duplicate keys is something that it says you "shouldnt" do. Even though they define "should", it doesn't really help from an implementation POV.

Out there, in reality, almost all of the major frameworks with JSON claim that "you shouldn't rely on key order" and they also don't duplicate keys. This is probably just a result of them using the lowest common denominator (if they produce JSON that relies on key ordering, and some library framework maps that `JObject` to a HashMap, then they can't use the library properly).

The other area of spec which isn't helpful is how it defines numbers

Regarding whether its a legal JSON string, I believe that any `JValue` should be completely legal for a default AST for a user. This is way `fast.JValue` was renamed to `unsafe.JValue`

> immutable vs mutable representations

Scala is idiomatically immutable, people tend to only use mutable for performance reasons. So for the user facing AST, I believe immutable is the way and what Scala users expect.

> allow or not allow duplicate keys

Honestly I think this should be a "no". It creates so many problems in regards to design that I don't think we will get anywhere. There is a legitimate reason to worry about key ordering (i.e. seeing if 2 JSON's are the same), but this one I believe should be firmly in the "don't take it into account" bucket.

Note that the current `unsafe.JValue` does support duplicate keys but its just a consequence of the data structures being used, its not intentional. We shouldn't expose the user facing AST with this functionality.

**Comment by mdedetrich** [comment]
Closing this issue, if there are any other comments regarding goals then just make a seperate issue for it
